yarn-project/prover-client/src/orchestrator/orchestrator.ts

```diff
@@ -0,0 +1,515 @@
+    currentLevel: bigint,
+    currentIndex: bigint,
+    mergeInputData: [BaseOrMergeRollupPublicInputs, Proof],
+    stateIdentifier: string,
+  ) {
+    const result = this.storeMergeInputs(currentLevel, currentIndex, mergeInputData);
+
+    // Are we ready to execute the next circuit?
+    if (!result.ready) {
+      return;
+    }
+
+    if (result.mergeLevel === 0n) {
+      this.checkAndExecuteRootRollup(stateIdentifier);
+    } else {
+      // onto the next merge level
+      this.enqueueJob(stateIdentifier, PROVING_JOB_TYPE.MERGE_ROLLUP, () =>
+        this.runMergeRollup(result.mergeLevel, result.indexWithinMergeLevel, result.mergeInputData, stateIdentifier),
+      );
+    }
+  }
+
+  /**
+   * Process the job queue
+   * Works by managing an input queue of proof requests and an active pool of proving 'jobs'
+   */
+  private async processJobQueue() {
+    // This is temporary to keep benchmarks consistent
+    const maxConcurrentJobs = 1;
+    // Used for determining the current state of a proving job
+    const promiseState = (p: Promise<void>) => {
+      const t = {};
+      return Promise.race([p, t]).then(
+        v => (v === t ? 'pending' : 'fulfilled'),
+        () => 'rejected',
+      );
+    };
+
+    // Just a short break between managing the sets of requests and active jobs
+    const createSleepPromise = () =>
+      sleep(SLEEP_TIME).then(_ => {
+        return 0;
+      });
+
+    let sleepPromise = createSleepPromise();
+    let promises: Promise<void>[] = [];
+    while (!this.stopped) {
+      // first look for more work
+      if (this.jobQueue.length() && promises.length < maxConcurrentJobs) {
+        // more work could be available
+        const job = await this.jobQueue.get();
+        if (job !== null) {
+          // a proving job, add it to the pool of outstanding jobs
+          promises.push(job.operation());
+        }
+        // continue adding more work
+        continue;
+      }
+
+      // no more work to add, here we wait for any outstanding jobs to finish and/or sleep a little
+      try {
+        const ops = Promise.race(promises).then(_ => {
+          return 1;
+        });
+        const result = await Promise.race([sleepPromise, ops]);
+        if (result === 0) {
+          // this is the sleep promise
+          // we simply setup the promise again and go round the loop checking for more work
+          sleepPromise = createSleepPromise();
+          continue;
+        }
+      } catch (err) {
+        // We shouldn't get here as all jobs should be wrapped in a 'safeJob' meaning they don't fail!
+        logger.error(`Unexpected error in proving orchestrator ${err}`);
+      }
+
+      // one or more of the jobs completed, remove them
+      const pendingPromises = [];
+      for (const jobPromise of promises) {
+        const state = await promiseState(jobPromise);
+        if (state === 'pending') {
+          pendingPromises.push(jobPromise);
+        }
+      }
```

Comment:
I looked around for alternative implementations to a promise pool so we didn't have to manually check the state of each promise, and found [this approach](https://github.com/rxaviers/async-pool/blob/master/lib/es9.js) which feels cleaner: they wrap each promise into a promise that returns itself, so it can be removed from the set when it finishes.